### PR TITLE
[RTM] Code revision: confounds, utilities, etc.

### PR DESCRIPTION
**Pull request by oesteban**
#225, depends on #640

### CODE REVIEW

File: fmriprep/interfaces/confounds.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+class GatherConfoundsInputSpec(BaseInterfaceInputSpec):
+    signals = File(exists=True, mandatory=True, desc='input signals')
+    dvars = File(exists=True, mandatory=True, desc='file containing DVARS')
+    fd = File(exists=True, mandatory=True, desc='input framewise displacement')
+    tcompcor = File(exists=True, mandatory=True, desc='input tCompCorr')
+    acompcor = File(exists=True, mandatory=True, desc='input aCompCorr')
+    cos_basis = File(exists=True, mandatory=True, desc='input cosine basis')
+    motion = File(exists=True, mandatory=True, desc='input motion parameters')
+    aroma = File(exists=True, mandatory=True, desc='input ICA-AROMA')
```
Comment (effigies) [inline fmriprep/interfaces/confounds.py]: These are not mandatory as presently written. I believe the idea is that the individual inputs can fail or (in the case of `aroma`, at least) not be run, and the interface will still work.

File: fmriprep/cli/run.py [hunk 2]
```diff
@@ -5,7 +5,6 @@
 fMRI preprocessing workflow
 =====
 """
-from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (effigies) [inline fmriprep/cli/run.py]: Is the plan to completely discontinue Py2 support? At least a couple months ago, we were accepting patches from users who needed Py2, as long as they didn't complicate the code.

File: fmriprep/interfaces/utils.py [hunk 3]
```diff
@@ -34,13 +41,100 @@ def _run_interface(self, runtime):
...
-    import nibabel as nb
-    import scipy.ndimage as nd
-    from nilearn.image import resample_to_img
+class TPM2ROIInputSpec(BaseInterfaceInputSpec):
+    in_file = File(exists=True, mandatory=True, desc='input file')
+    in_mask = File(exists=True, mandatory=True, desc='input mask')
+    mask_erode_mm = traits.Float(0.0, usedefault=True,
+                                 desc='erode input mask (kernel width in mm)')
+    erode_mm = traits.Float(0.0, usedefault=True,
+                            desc='erode output mask (kernel width in mm)')
```
Comment (effigies) [inline fmriprep/interfaces/utils.py]: While we're making this interface, let's put in a probability threshold parameter (default 0.95).

File: fmriprep/utils/misc.py [hunk 4]
```diff
@@ -2,52 +2,39 @@
 # -*- coding: utf-8 -*-
 # emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
 # vi: set ft=python sts=4 ts=4 sw=4 et:
-from __future__ import print_function, division, absolute_import, unicode_literals
+"""
+Miscelaneous utilities
```
Comment (effigies) [inline fmriprep/utils/misc.py]: Miscellaneous

File: fmriprep/utils/misc.py [hunk 5]
```diff
@@ -2,52 +2,39 @@
...
+
+    >>> add_suffix([
+    ...     '/path/to/sub-045_ses-test_T1w.nii.gz',
+    ...     '/path/to/sub-045_ses-retest_T1w.nii.gz'], 'test')
+    'sub-045_ses-test_T1w_test.nii.gz'
+
+    """
     import os.path as op
     from niworkflows.nipype.utils.filemanip import fname_presuffix, filename_to_list
     return op.basename(fname_presuffix(filename_to_list(in_files)[0],
```
Comment (effigies) [inline fmriprep/utils/misc.py]: To get the behavior described in your docstring: `suffix='_' + suffix`

Or update the docstring to pass `'_test'`.

File: fmriprep/cli/run.py (hunk 2 above)
Comment (oesteban) [inline fmriprep/cli/run.py]: I would discontinue py2 for good at this point, simplify our lives. But you are right, we haven't talked about this. WDYT @chrisfilo @rwblair ?

File: fmriprep/interfaces/utils.py (hunk 3 above)
Comment (oesteban) [inline fmriprep/interfaces/utils.py]: Done

File: fmriprep/utils/misc.py (hunk 5 above)
Comment (oesteban) [inline fmriprep/utils/misc.py]: Done

File: fmriprep/interfaces/fmap.py [hunk 6]
```diff
@@ -138,11 +136,11 @@ def _despike2d(data, thres, neigh=None):
...
     nslices = data.shape[-1]
 
-    for k in range(nslices):
+    for k in list(range(nslices)):
         data2d = data[..., k]
 
-        for i in range(data2d.shape[0]):
-            for j in range(data2d.shape[1]):
+        for i in list(range(data2d.shape[0])):
+            for j in list(range(data2d.shape[1])):
```
Comment (effigies) [inline fmriprep/interfaces/fmap.py]: Why `list`? You can iterate over `range`.

File: fmriprep/interfaces/images.py [hunk 7]
```diff
@@ -77,8 +75,8 @@ def _run_interface(self, runtime):
             in_files = [self.inputs.in_files]
 
         # Generate output average name early
-        self._results['out_avg'] = genfname(self.inputs.in_files[0],
-                                            suffix='avg')
+        self._results['out_avg'] = fname_presuffix(self.inputs.in_files[0],
+                                                   suffix='avg', newpath=os.getcwd())
```
Comment (effigies) [inline fmriprep/interfaces/images.py]: `suffix='_avg', newpath=runtime.cwd`

File: fmriprep/interfaces/images.py [hunk 8]
```diff
@@ -93,7 +91,8 @@ def _run_interface(self, runtime):
                 if sqdata.ndim == 5:
                     raise RuntimeError('Input image (%s) is 5D' % in_files[0])
                 else:
-                    in_files = [genfname(in_files[0], suffix='squeezed')]
+                    in_files = [fname_presuffix(in_files[0], suffix='squeezed',
+                                                newpath=os.getcwd())]
```
Comment (effigies) [inline fmriprep/interfaces/images.py]: `suffix='_squeezed', newpath=runtime.cwd`

File: fmriprep/interfaces/itk.py [hunk 9]
```diff
@@ -75,8 +75,8 @@ def _run_interface(self, runtime):
         field = field[:, :, :, np.newaxis, :]
 
         # Write out
-        self._results['out_file'] = genfname(
-            self.inputs.in_file, suffix='antswarp')
+        self._results['out_file'] = fname_presuffix(
+            self.inputs.in_file, suffix='antswarp', newpath=os.getcwd())
```
Comment (effigies) [inline fmriprep/interfaces/itk.py]: etc.

File: fmriprep/interfaces/utils.py [hunk 10]
```diff
@@ -34,23 +40,113 @@ def _run_interface(self, runtime):
...
+    pthres = traits.Float(0.95, usedefault=True,
+                          desc='threshold for the tissue probability maps')
```
Comment (effigies) [inline fmriprep/interfaces/utils.py]: I'd prefer this be something like `probability_threshold`, or `prob_thresh` or something a little more explicit. The main place we'll be reading it is

```Python
node = pe.Node(TPM2ROI(probability_threshold=0.95), ...)
```

File: fmriprep/utils/bids.py [hunk 11]
```diff
@@ -48,10 +65,13 @@ def collect_participants(bids_dir, participant_label=None, strict=False):
...
     >>> collect_participants('ds114', participant_label=['02', '14'],
-    ...                      strict=True)
+    ...                      strict=True)  # doctest: +IGNORE_EXCEPTION_DETAIL
     Traceback (most recent call last):
+    fmriprep.utils.bids.BIDSError:
     ...
-    RuntimeError: Some participants were not found: 14
+
+
+
```
Comment (effigies) [inline fmriprep/utils/bids.py]: Seems like more newlines than we need?

**Comment by oesteban** [comment]
Testing documentation: https://readthedocs.org/projects/fmriprep/builds/5792532/

CircleCI: https://circleci.com/gh/poldracklab/fmriprep/1006

**Comment by oesteban** [comment]
Tests are passing here: https://circleci.com/gh/poldracklab/fmriprep/1006 (the circle hook is showing a branch I created to test that it was building on rtd). Merging in...

**Pull request merged**